src/Microsoft.Azure.SignalR.Common/ClientInvocation/ClientResultsManager.cs

```diff
@@ -0,0 +1,312 @@
+        {
+            if (_routedInvocations.TryGetValue(message.InvocationId!, out var item))
+            {
+                if (item.ConnectionId != connectionId)
+                {
+                    throw new InvalidOperationException($"Connection ID '{connectionId}' is not valid for invocation ID '{message.InvocationId}'.");
+                }
+
+                // if false the connection disconnected right after the above TryGetValue
+                // or someone else completed the invocation (likely a bad client)
+                // we'll ignore both cases
+                if (_routedInvocations.Remove(message.InvocationId!, out _))
+                {
+                    item.Complete(item.Tcs, message);
+                }
+            }
+            else
+            {
+                // connection was disconnected or someone else completed the invocation
+            }
+        }
+
+        public (Type Type, string ConnectionId, object Tcs, Action<object, CompletionMessage> Completion)? RemoveInvocation(string invocationId)
+        {
+            _pendingInvocations.TryRemove(invocationId, out var item);
+            return item;
+        }
+
+        public (Type Type, string ConnectionId, string callerServerId, object Tcs, Action<object, CompletionMessage> Completion)? RemoveRoutedInvocation(string invocationId)
+        {
+            _routedInvocations.TryRemove(invocationId, out var item);
+            return item;
+        }
+
+        public Type GetReturnType(string invocationId)
+        {
+            if (TryGetType(invocationId, out var type))
+            {
+                return type;
+            }
+            throw new InvalidOperationException($"Invocation ID '{invocationId}' is not associated with a pending client result.");
+        }
+
+        public bool TryGetType(string invocationId, out Type type)
+        {
+            if (_pendingInvocations.TryGetValue(invocationId, out var item1))
+            {
+                type = item1.Type;
+                return true;
+            }
+            if (_routedInvocations.TryGetValue(invocationId, out var item2))
+            {
+                type = item2.Type;
+                return true;
+            }
+            type = null;
+            return false;
+        }
+
+        public IReadOnlyList<Type> GetParameterTypes(string methodName)
+        {
+            throw new NotImplementedException();
+        }
+
+        // Unused, here to honor the IInvocationBinder interface but should never be called
+        public Type GetStreamItemType(string streamId)
+        {
+            throw new NotImplementedException();
+        }
+
+        // Custom TCS type to avoid the extra allocation that would be introduced if we managed the cancellation separately
+        // Also makes it easier to keep track of the CancellationTokenRegistration for disposal
+        internal sealed class TaskCompletionSourceWithCancellation<T> : TaskCompletionSource<T>
```

Comment:
not quite get what this class is for

Reply:
See [this comment](https://github.com/Azure/azure-signalr/pull/1684/files/7a1316985004302c680a5519cec9ea85d97950d9#r979471825)